cy.selectFile docs (#4241)
cypress-io/cypress-documentation: content/api/commands/selectfile.md

---
title: selectFile
---

> Attaches a file or files to an HTML5 input element or simulates dragging a file
> or files into the browser.

Reviewer: ```suggestion
Selects a file or files to be uploaded to an HTML5 input element or simulates dragging a file
or files into the browser.
```

Author: Going for a slightly different edit here:

> Selects a file or files to attach to an HTML5 input element or simulates dragging a file or files into the browser.

I don't necessarily think we need to avoid the word attach, since it's still more descriptive of what's actually happening than 'upload' is.

Reviewer: It's replacing 'attach' with 'to be uploaded' not directly the verb 'upload' since it's not uploaded yet. I think upload is more descriptive. If you look at the docs we sighted, like mdn - it uses attach 2 times and upload 35 times for instance. When I search for this feature in google I always type 'upload files'. I don't think attach is more descriptive than upload. @cowboy thoughts?

## Syntax

```javascript
.selectFile(file)
.selectFile(file, options)
.selectFile([file1, file2, ...])
.selectFile([file1, file2, ...], options)
```

### Usage

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```javascript
cy.get('input[type=file]').selectFile('file.json')
cy.get('input[type=file]').selectFile(['file.json', 'file2.json'])
cy.get('input[type=file]').selectFile({

>   contents: Buffer.from('foo'),

Reviewer 2: Can we change `foo` to `file contents` to better mirror the "incorrect usage" example below? And also change `file.json` to something like `file.txt` because either of the above are invalid JSON

Author: Done and done.

  fileName: 'file.json',
  lastModified: Date.now(),
})

> ```

Reviewer 2: Can we show one or more drag+drop examples here, eg.

```js
cy.get('.file-target').selectFile('file.json', { action: 'drag-drop' })
cy.document().selectFile('file.json', { action: 'drag-drop' })
```

Author: Added.

**<Icon name="exclamation-triangle" color="red"></Icon> Incorrect Usage**

```javascript
// Errors, cannot be chained off 'cy'
cy.selectFile('file.json')

// Will attempt to find a file called 'file contents'
// on disk, probably not what you intended
cy.get('input[type=file]').selectFile('file contents')
```

### Arguments

**<Icon name="angle-right"></Icon> file** **_(String, Array, Object or
Buffer)_**

Either a single file, or an array of files. A file can be:

- A path to a file within the project root (the directory that contains the
  default Cypress configuration file). Eg: `'path/to/file.json'`
- `@alias` - An alias of any type, previously stored using `.as()`. Eg:
  `'@alias'`

> - A `Buffer()` containing binary data, such as that returned by

Reviewer 2: Same question here as in the migration guide WRT `Buffer`. Are people going to understand what a `Buffer` is? Do we need to link to something here or explain this below?

Author: Added the same link to the node docs here.

Reviewer 2: Does cypress provide `Buffer` for people to use, or is it something they will need to import in their spec file? It may be useful to show that in an example on this page, because it doesn't seem obvious (like, I personally do not know what I would need to do to use `Buffer` like in these examples, and would have to google it)

Author: Buffer is a global glass, same as Array or Object. I don't think there's anything more that needs to be done to make this explicit - the examples are complete. There's no import statement needed. If they copy paste the examples exactly as written, they will just work.

We don't explicitly call out that `cy` is provided as a global either - we just use it in our examples and assume people will understand and try it. :shrug:

Reviewer 2: Ok, cool. As long as it works!

Reviewer 2: I just ran a few tests with this branch. When using component testing, projects that use Webpack will polyfill `Buffer`, so your examples work as-is. However, projects that use Vite do not polyfill `Buffer`, so we end up seeing a `Buffer is not defined` error, eg.

![image](https://user-images.githubusercontent.com/54051/148593663-58f5db22-90df-4206-9a8d-d83064697b65.png)

We can work around this issue by polyfilling `Buffer` with the [buffer](https://www.npmjs.com/package/buffer) npm module (per this issue https://github.com/vitejs/vite/discussions/2785) and then doing something like this at the top of any spec files that use `Buffer`:

```js
import { Buffer } from 'buffer'
```

or globally in the `supportFile`:

```js
import { Buffer } from 'buffer'
globalThis.Buffer = Buffer
```

We should add a note to this effect somewhere!

Reviewer 2: @BlueWinds I'm happy to take a stab at writing this, just LMK and I'll make a commit.

Author: Huh, today I learned that we ship with a Vite based dev server! https://docs.cypress.io/guides/component-testing/framework-configuration#Vite-Based-Projects-Vue-React

I'm getting set up with a vite project now - ideally we'd solve this in our code rather than our documentation, so that users have a consistent experience. Now I'm wondering if the readFile option to use Buffers works there at all, or if it's been broken since release and no one has noticed.

Author: The link you provided suggests we should be able to do `globalThis.Buffer = Buffer` just once, and it should work everywhere (including in user's spec file).

Reviewer 2: If you need examples, also see:

- https://github.com/cypress-io/cypress-component-testing-examples/tree/main/vite-react
- https://github.com/cypress-io/cypress-component-testing-examples/tree/main/vite-vue

| `fileName`     | The name of the file. If `contents` is a path on disk, this defaults to the actual filename. In any other case, this defaults to an empty string.                                                                                                  |
| `lastModified` | The file's last modified timestamp, in milliseconds elapsed since the UNIX epoch (eg. [`Date.prototype.getTime()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)). This defaults to `Date.now()`. |

**<Icon name="angle-right"></Icon> options** **_(Object)_**

Pass in an options object to change the default behavior of `.selectFile()`.

| Option                       | Default                                                                        | Description                                                                                                                                       |
| ---------------------------- | ------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `action`                     | `'select'`                                                                     | Switches selectFile between modes. Valid values are `select` and `drag-drop`. See [Actions](#Actions) below for more details.                     |

Reviewer 2: Can you either format selectFile as `.selectFile()` here or remove "selectFile between"?

Author: Removed extra words.

| `log`                        | `true`                                                                         | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log).                                                         |
| `timeout`                    | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)           | Time to wait for `.selectFile()` to resolve before [timing out](#Timeouts).                                                                       |
| `waitForAnimations`          | [`waitForAnimations`](/guides/references/configuration#Actionability)          | Whether to wait for elements to [finish animating](/guides/core-concepts/interacting-with-elements#Animations) before executing the command.      |

### Yields [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management)

- `.selectFile()` yields the same subject it was given from the previous
  command.

### Action

Depending on the `action` set in the `options` argument, `.selectFile()` can
simulate two different user behaviors:

#### `select` (default)

> By default, `selectFile()` runs in 'select' mode, mimicking a user selecting one

Reviewer 2: ```suggestion
By default, `.selectFile()` runs in 'select' mode, mimicking a user selecting one
```

Author: Applied as part of other commits.

or more files on an HTML5 input element. In this mode, the
[subject](/guides/core-concepts/introduction-to-cypress#Subject-Management) must
be a single `input` element with `type="file"`, or a `label` element connected
to an input (either with its `for` attribute or by containing the input).

#### `drag-drop`

Setting the action to `drag-drop` changes the behavior of the command to instead
mimic a user dragging files from the operating system into the browser, and
dropping them over the selected subject. In this mode, the subject can be any
DOM element or the `document` as a whole.

## Examples

### From a file on disk

```javascript
cy.get('input[type=file]').selectFile('path/to/file.json')
cy.get('input[type=file]').selectFile('path/to/file.png')
```

> If given a path, `.selectFile()` will search for the file relative to the
> project root and attach the file exactly as it exists on disk. This is the
> preferred way to work with files on disk, avoiding many encoding-related
> pitfalls.

Reviewer: ```suggestion
If given a path, `.selectFile()` will search for the file relative to the
project root and upload the file exactly as it exists on disk. This is the
preferred way to work with files on disk, avoiding many encoding-related
pitfalls.
```

Author: Going to leave this one as-is, since as mentioned above, I think attach is more descriptive than upload in this context.

### On a hidden input

```javascript
cy.get('input[type=file]').selectFile('file.json', { force: true })
```

> In many cases in modern applications, the underlying file input is hidden from
> view, and activated by a user clicking on a button. In these cases, you will
> need to tell Cypress to ignore its actionability checks and attach the file even
> though a user would not be able to directly activate the file input.

Reviewer: ```suggestion
In many cases in modern applications, the underlying file input is hidden from
view, and activated by a user clicking on a button. In these cases, you will
need to tell Cypress to ignore its actionability checks and upload the file even
though a user would not be able to directly activate the file input.
```

Author: Going to use 'select' rather than 'attach' here.

### From a fixture

```javascript
cy.fixture('file.json', { encoding: null }).as('myFixture')
cy.get('input[type=file]').selectFile('@myFixture')
```

Note the use of `null` encoding. By default, `cy.fixture()` and `cy.readFile()`
attempt to interpret files read from disk, which would result in a JSON file
being decoded and re-encoded as a utf-8 string - the contents would be
preserved, but formatting would not be and the encoding might change. See
[`cy.fixture`](/api/commands/fixture) or [`cy.readFile`](/api/commands/readfile)
for more details on file encoding.

### From an API response

```javascript
cy.request('http://localhost:8888/users/827').its('body').as('responseBody')

cy.get('input[type=file]').selectFile('@responseBody')
```

### Processing data inside the test

```javascript
cy.readFile('users.json')
  .then((users) => {
    users[0].username = 'JohnCena'
  })
  .as('myFile')

cy.get('input[type=file]').selectFile('@myFile')
```

> ### Attaching multiple files

Reviewer: ```suggestion
### Selecting multiple files
```

```javascript
cy.get('input[type=file]').selectFile([
  'file1.json',
  'file2.json',
  'file3.json',
])
```

<Alert type="warning">

This will fail unless the file input has the `multiple` property.

</Alert>

> ### Attaching a file with custom filename and lastModified

Reviewer: ```suggestion
### Selecting a file with custom filename and lastModified
```

```javascript
cy.get('input[type=file]').selectFile({
  contents: 'path/to/file.json',

>   fileName: 'users.json',

Reviewer 2: Maybe this can be more clear like so:

```suggestion
  fileName: 'custom-name.json',
```

Author: Done.

## Rules

### Requirements [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Chains-of-Commands)

- `.selectFile()` requires being chained off a command that yields DOM
  element(s). With the `input` action (default), it further requires a single
  `input` element with `type="file"`, or a `label` element attached to one.
- If given a path, `.selectFile()` requires the file must exist.
- If given an alias, `.selectFile()` requires that the subject of the alias must
  not be `null` or `undefined`.

### Assertions [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Assertions)

- `.selectFile()` will automatically wait for the element to reach an
  [actionable state](/guides/core-concepts/interacting-with-elements).

### Timeouts [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Timeouts)

- `.selectFile()` can time out waiting for the element to reach an
  [actionable state](/guides/core-concepts/interacting-with-elements).
- `.selectFile()` can time out waiting for a file to exist on disk or for an
  alias to resolve.

## Command Log

> **_Attach file to input_**

Reviewer: ```suggestion
**_Select file to upload to input_**
```

Author: Applying slightly different version of this one

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```javascript
cy.get('input[type=file]').selectFile('file.json')
cy.get('input[type=file]').selectFile(['file.json', 'file2.json'])

cy.get('input[type=file]').selectFile({
  contents: Cypress.Buffer.from('file contents'),
  fileName: 'file.txt',
  lastModified: Date.now(),
})

cy.get('input[type=file]').selectFile('file.json', { action: 'drag-drop' })
cy.document().selectFile('file.json', { action: 'drag-drop' })
```

**<Icon name="exclamation-triangle" color="red"></Icon> Incorrect Usage**

```javascript
// Errors, cannot be chained off 'cy'
cy.selectFile('file.json')

// Will attempt to find a file called 'file contents'
// on disk, probably not what you intended
cy.get('input[type=file]').selectFile('file contents')
```

### Arguments

**<Icon name="angle-right"></Icon> file** **_(String, Array, Object or
Cypress.Buffer)_**

Either a single file, or an array of files. A file can be:

- A path to a file within the project root (the directory that contains the
  default Cypress configuration file). Eg: `'path/to/file.json'`
- `@alias` - An alias of any type, previously stored using `.as()`. Eg:
  `'@alias'`

> - A [`Cypress.Buffer()`](/api/utilities/buffer) containing binary data, such as

Reviewer 2: I don't think we want to show `Cypress.Buffer` being called as a function. How about this?

```suggestion
- A [`Cypress.Buffer`](/api/utilities/buffer) instance containing binary data, such as
```

  that returned by `cy.readFile('file.json', { encoding: null })`. Eg:
  `Cypress.Buffer.from('foo')`
- An object with a non-null `contents` property, specifying details about the
  file. Eg: `{contents: '@alias', fileName: 'file.json'}`

If an object is provided, it can have the following properties.

| Option         | Description                                                                                                                                                                                                                                        |
| -------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `contents`     | The contents of the file. This can be a string shorthand as described above, a `Cypress.Buffer()` containing binary data or a non-Buffer object, which will be converted into a string with `JSON.stringify()` and `utf8` encoded.                 |

Reviewer 2: ```suggestion
| `contents`     | The contents of the file. This can be a string shorthand as described above, a `Cypress.Buffer` instance containing binary data or a non-Buffer object, which will be converted into a string with `JSON.stringify()` and `utf8` encoded.                 |
```
